ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -911,45 +952,80 @@ static bool ShouldHandleFileOpEvent(
         return false;
     }
     
+    bool shouldHandle = false;
+    
     {
         PerfSample findRootSample(perfTracer, PrjFSPerfCounter_FileOp_ShouldHandle_FindVirtualizationRoot);
         
-        *root = VirtualizationRoot_FindForVnode(
+        *vnodeRoot = VirtualizationRoot_FindForVnode(
             perfTracer,
             PrjFSPerfCounter_FileOp_FindRoot,
             PrjFSPerfCounter_FileOp_FindRoot_Iteration,
             vnode,
             context);
         
-        if (!VirtualizationRoot_IsValidRootHandle(*root))
+        if (VirtualizationRoot_IsValidRootHandle(*vnodeRoot))
+        {
+            *pid = GetPid(context);
+            pidIsValid = true;
+            
+            shouldHandle = ShouldMessageProviderWithHandle(perfTracer, *vnodeRoot, *pid);
+        }
+        else
         {
             perfTracer->IncrementCount(PrjFSPerfCounter_FileOp_ShouldHandle_NoRootFound);
-            return false;
         }
     }
     
+    if (fromPath != nullptr)
     {
-        PerfSample checkRootOnlineSample(perfTracer, PrjFSPerfCounter_FileOp_ShouldHandle_CheckProvider);
-        
-        ActiveProviderProperties provider = VirtualizationRoot_GetActiveProvider(*root);
-        
-        if (!provider.isOnline)
+        VirtualizationRootHandle fromRoot = ActiveProvider_FindForPath(fromPath);
+        if (VirtualizationRoot_IsValidRootHandle(fromRoot))
         {
-            perfTracer->IncrementCount(PrjFSPerfCounter_FileOp_ShouldHandle_OfflineRoot);
-            return false;
+            if (!pidIsValid)
+            {
+                *pid = GetPid(context);
+            }
+            
+            shouldHandleFromProvider = ShouldMessageProviderWithHandle(perfTracer, fromRoot, *pid);
+            if (shouldHandleFromProvider)
+            {
+                *fromProvider = fromRoot;
+            }
         }
-    
-        // If the calling process is the provider, we must exit right away to avoid deadlocks
-        *pid = GetPid(context);
-        if (*pid == provider.pid)
+        else
         {
-            perfTracer->IncrementCount(PrjFSPerfCounter_FileOp_ShouldHandle_OriginatedByProvider);
-            return false;
+            perfTracer->IncrementCount(PrjFSPerfCounter_FileOp_ShouldHandle_NoRootFound);
         }
     }
+    
+    if (shouldHandle || shouldHandleFromProvider)
+    {
+        *vnodeFsidInode = Vnode_GetFsidAndInode(vnode, context, true /* the inode is used for getting the path in the provider, so use linkid */);
+    }
 
-    *vnodeFsidInode = Vnode_GetFsidAndInode(vnode, context, true /* the inode is used for getting the path in the provider, so use linkid */);
+    return shouldHandle;
```

Comment:
If you check the relevant rename/link call site for this in `HandleFileOpOperation`, you'll find that the return value is treated as wether "we should message the destination provider":

```
bool messageDestinationProvider = ShouldHandleFileOpEvent(
```

whereas "we should message the source provider" is determined by whether we got a valid handle for the "from" root.

```
bool messageSourceProvider = VirtualizationRoot_IsValidRootHandle(fromRoot);
```

We could certainly return `shouldHandle || shouldHandleFromProvider` but then we need to additionally convey whether or not to message the destination provider.

I think the calling semantics of `ShouldHandleFileOpEvent` are awful (hence #900) and this PR certainly makes things worse rather than better. Fixing them is out of scope here, and I'm not sure what would be a better approach as a stopgap. Returning a struct of 2 booleans is going to make things worse for all the 1-root call sites, and your proposed return value would imply we need to either change the semantics of `VirtualizationRootHandle* vnodeRoot` so that a `None` handle is returned if the destination provider *shouldn't* be messaged, (I guess this should be safe, and our now-excellent unit test coverage should tell us if not?) or we need an extra output argument.

What's your preferred option?